### PR TITLE
Add Conjugate Gradient example to benchmarks

### PR DESCRIPTION
**Pull request by shehzan10**
Compare the performance and memory usage of sparse vs dense using conjugate
gradient example

[skip arrayfire ci]

### CODE REVIEW

File: examples/benchmarks/cg.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+using namespace af;
+
+static size_t dimension = 4 * 1024;
+static const int maxIter = 10;
+static const int sparsityFactor = 7;
+
+static array A;
+static array spA;  // Sparse A
+static array x0;
```
Comment (umar456) [inline examples/benchmarks/cg.cpp]: this looks too similar to a hex number.

File: examples/benchmarks/cg.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+}
+
+void denseConjugateGradient(void)
+{
+    array x = constant(0, b.dims(), f32);
+    array r = b - matmul(A, x);
+    array p = r;
+
+    for (int i = 0; i < maxIter; ++i) {
```
Comment (umar456) [inline examples/benchmarks/cg.cpp]: doesn't the timeit function execute this multiple times anyway?

File: examples/benchmarks/cg.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        array Ap = matmul(in, p);
+        array alpha_num = dot(r, r);
+        array alpha_den = dot(p, Ap);
+        array alpha = alpha_num/alpha_den;
+        r = r - tile(alpha, Ap.dims())*Ap;
+        x = x + tile(alpha, Ap.dims())*p;
+        array beta_num = dot(r, r);
+        array beta = beta_num/alpha_num;
+        p = r + tile(beta, p.dims())*p;
```
Comment (umar456) [inline examples/benchmarks/cg.cpp]: `p *= r + tile(beta, p.dims());`

File: examples/benchmarks/cg.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    array r = b - matmul(in, x);
+    array p = r;
+
+    for (int i = 0; i < maxIter; ++i) {
+        array Ap = matmul(in, p);
+        array alpha_num = dot(r, r);
+        array alpha_den = dot(p, Ap);
+        array alpha = alpha_num/alpha_den;
+        r = r - tile(alpha, Ap.dims())*Ap;
+        x = x + tile(alpha, Ap.dims())*p;
```
Comment (umar456) [inline examples/benchmarks/cg.cpp]: `x += tile(alpha, Ap.dims())*p;`

File: examples/benchmarks/cg.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    array x = constant(0, b.dims(), f32);
+    array r = b - matmul(in, x);
+    array p = r;
+
+    for (int i = 0; i < maxIter; ++i) {
+        array Ap = matmul(in, p);
+        array alpha_num = dot(r, r);
+        array alpha_den = dot(p, Ap);
+        array alpha = alpha_num/alpha_den;
+        r = r - tile(alpha, Ap.dims())*Ap;
```
Comment (umar456) [inline examples/benchmarks/cg.cpp]: r -= tile(alpha, Ap.dims())*Ap;

File: examples/benchmarks/cg.cpp (hunk 1 above)
Comment (shehzan10) [inline examples/benchmarks/cg.cpp]: x0 is a mathematical notation used in conjugate gradient.

File: examples/benchmarks/cg.cpp (hunk 2 above)
Comment (shehzan10) [inline examples/benchmarks/cg.cpp]: This is not for timing, conjugate gradient is an iterative solver.

Comment (umar456) [inline examples/benchmarks/cg.cpp]: Right. I don't know what I was thinking.

File: examples/benchmarks/cg.cpp (hunk 3 above)
Comment (kaatish) [inline examples/benchmarks/cg.cpp]: @umar456 These expressions are not equivalent. We want something like this `p = r + scalar*p`.

Comment (umar456) [inline examples/benchmarks/cg.cpp]: doh! Clearly its time for bed.

File: examples/benchmarks/cg.cpp (hunk 1 above)
Comment (pavanky) [inline examples/benchmarks/cg.cpp]: @shehzan10 FYI, try to use X0 in such cases.

**Pull request merged**